esp-hal/src/dma/mod.rs

```diff
@@ -732,59 +730,67 @@ where
+            if self.last_seen_handled_descriptor_ptr.is_null() {
+                self.last_seen_handled_descriptor_ptr = last_in_dscr_ptr;
+                self.read_buffer_start = last_in_descr.buffer;
+                self.available = len;
+            } else if self.last_seen_handled_descriptor_ptr != last_in_dscr_ptr {
+                self.available += len;
+            }
         }
 
-        let descr_address = self.last_seen_handled_descriptor_ptr;
-        let mut dw0 = unsafe { descr_address.read_volatile() };
+        self.last_seen_handled_descriptor_ptr = last_in_dscr_ptr;
+        self.available
+    }
 
-        if dw0.owner() == Owner::Cpu && !dw0.is_empty() {
-            let descriptor_buffer = dw0.buffer;
-            let next_descriptor = dw0.next;
+    fn pop(&mut self, data: &mut [u8]) -> Result<usize, DmaError> {
+        let mut avail = self.available;
 
-            self.read_buffer_start = descriptor_buffer;
-            self.available = dw0.len();
+        let mut remaining = data.len();
+        let mut offset = 0;
+        let mut descr_ptr = self.last_seen_handled_descriptor_ptr;
 
-            dw0.set_owner(Owner::Dma);
-            dw0.set_length(0);
-            dw0.set_suc_eof(false);
+        if descr_ptr.is_null() {
+            return Ok(0);
+        }
 
+        while remaining != 0 && remaining >= avail {
             unsafe {
-                descr_address.write_volatile(dw0);
+                let mut descr = descr_ptr.read_volatile();
+                let dst = data.as_mut_ptr().add(offset);
+                let src = descr.buffer;
+                let count = descr.len();
+                core::ptr::copy_nonoverlapping(src, dst, count);
+
+                descr.set_owner(Owner::Dma);
+                descr.set_suc_eof(false);
+                descr.set_length(0);
+                descr_ptr.write_volatile(descr);
+
+                remaining -= count;
+                offset += count;
+                avail -= count;
+                descr_ptr = descr.next;
             }
 
-            self.last_seen_handled_descriptor_ptr = if next_descriptor.is_null() {
-                self.descriptors.as_mut_ptr()
-            } else {
-                next_descriptor
-            };
-        }
-
-        self.available
-    }
-
-    fn pop(&mut self, data: &mut [u8]) -> Result<usize, DmaError> {
-        let avail = self.available;
-
-        if avail < data.len() {
-            return Err(DmaError::Exhausted);
+            if descr_ptr.is_null() {
+                break;
+            }
         }
 
-        unsafe {
-            let dst = data.as_mut_ptr();
-            let src = self.read_buffer_start;
-            let count = self.available;
-            core::ptr::copy_nonoverlapping(src, dst, count);
+        if avail > 0 {
+            return Err(DmaError::BufferTooSmall);
         }
 
-        self.available = 0;
-        Ok(data.len())
+        self.last_seen_handled_descriptor_ptr = core::ptr::null_mut();
```

Comment:
Should this be?
```suggestion
        self.last_seen_handled_descriptor_ptr = descr_ptr;
```

Or maybe not that exactly but something about setting this to null (when `self.available` is not 0) seems wrong but I can't put my finger on it.